src/components/MachineSelector/Configurator/MachineConfigurator.tsx

```diff
@@ -0,0 +1,337 @@
+  id: string;
+  children: React.ReactNode;
+  onClose: () => void;
+}
+
+/**
+ * This component is used as machine configurator and enabled only in settings page for admin users
+ */
+
+/**
+ * Title component for machine configuration popup
+ */
+const DialogTitle = withStyles(styles)((props: DialogTitleProps) => {
+  const { children, classes, onClose, ...other } = props;
+  return (
+    <MuiDialogTitle disableTypography className={classes.root} {...other}>
+      <Typography variant="h6">{children}</Typography>
+      {onClose ? (
+        <IconButton
+          aria-label="close"
+          className={classes.closeButton}
+          onClick={onClose}
+        >
+          <CloseIcon />
+        </IconButton>
+      ) : null}
+    </MuiDialogTitle>
+  );
+});
+
+type FormField = {
+  assetExternalId: string;
+};
+
+const defaultValues = {
+  assetExternalId: '',
+};
+
+/**
+ * This is the main functional component having the machine configurator
+ */
+const MachineConfigurator: FC<Props> = (props: Props) => {
+  const [open, setOpen] = useState(false);
+  const [assetExternalIds, setAssetExternalIds] = useState<string[]>([]);
+  const [assetValidating, setAssetValidating] = useState(false);
+
+  const onError = (msg: string) => {
+    props.setAlerts({
+      type: 'error',
+      text: msg,
+      hideApp: false,
+    });
+  };
+
+  const { handleSubmit, control, reset } = useForm({
+    defaultValues,
+  });
+
+  /**
+   * This will open the configurator popup on machine settings icon click.
+   * restoreMachineConfig function is used to fill the form fields with previously saved values
+   */
+  const handleClickOpen = () => {
+    setOpen(true);
+    setTimeout(() => {
+      restoreMachineConfig();
+    }, 10);
+  };
+
+  /**
+   * This will close the popup
+   */
+  const handleClose = () => {
+    setOpen(false);
+  };
+
+  /**
+   * This is used to update the form fields with previously saved values
+   */
+  const restoreMachineConfig = async () => {
+    const savedMachineConfigStr = await getMachineIds(onError);
+    if (savedMachineConfigStr) {
+      setAssetExternalIds(savedMachineConfigStr.split(','));
+    }
+  };
+
+  /**
+   * This function is used to validate machine saving.
+   */
+  const isValidSaving = () => {
+    if (assetExternalIds.length === 0) {
+      props.setAlerts({
+        type: 'warning',
+        text: 'Atleast one Asset External Id is required!',
+        hideApp: false,
+      });
+      return false;
+    }
+    return true;
+  };
+
+  /**
+   * This function is used to save machine configurations.
+   * A alert will be displayed finally based on the saving status.
+   */
+  const saveMachineConfig = async () => {
+    if (isValidSaving()) {
+      const actionStatus = await saveMachineIds(
+        assetExternalIds.join(','),
+        onError
+      );
+      if (actionStatus) {
+        handleClose();
+        props.updateAssets();
```

Comment:
What is the significant of this?

Reply:
Once we updated the asset external ids, state should be updated with the new asset(machine) list.